File: ui/src/common/types/channel/channel.ts
import type {
  ChannelId,
  ChannelSessionId,
  ChannelUserId,
  CompanionId,
  ConversationId,
  PublicAgentId,
} from '@/common/types/ids';

export interface IChannelPluginStatus {
  /** Canonical persisted channel entity id. */
  id: ChannelId;
  type: string;
  name: string;
  enabled: boolean;
  connected: boolean;
  status?: string;
  last_connected?: number;
  error?: string;
  activeUsers: number;
  botUsername?: string;
  hasToken?: boolean;
  /** 绑定的伙伴（每机器人一宠；UNIQUE(type,bot_key) 保证同一机器人不绑多宠）。 */
  companionId?: CompanionId;
  /** 绑定的对外伙伴（与 companionId 互斥；一个机器人只服务一个对象）。 */
  publicAgentId?: PublicAgentId | null;
  /** 平台级机器人身份（lark app_id / telegram bot id / ...）。 */
  botKey?: string;
  isExtension?: boolean;
  extensionMeta?: {
    credentialFields?: Array<{
      key: string;
      label: string;
      type: 'text' | 'password' | 'select' | 'number' | 'boolean';
      required?: boolean;
      options?: string[];
      default?: string | number | boolean;
    }>;
    configFields?: Array<{
      key: string;
      label: string;
      type: 'text' | 'password' | 'select' | 'number' | 'boolean';
      required?: boolean;
      options?: string[];
      default?: string | number | boolean;
    }>;
    description?: string;
    extensionName?: string;
    icon?: string;
  };
}

export interface IChannelPairingRequest {
  code: string;
  platformUserId: string;
  platformType: string;
  display_name?: string;
  requestedAt: number;
  expiresAt: number;
  /** 发起/归属的机器人渠道行 id；旧库未回填时可能缺省。 */
  channelId?: ChannelId;
}

export interface IChannelUser {
  id: ChannelUserId;
  platformUserId: string;
  platformType: string;
  display_name?: string;
  authorizedAt: number;
  lastActive?: number;
  session_id?: ChannelSessionId;
  /** 发起/归属的机器人渠道行 id。 */
  channelId?: ChannelId;
}

export interface IChannelSession {
  id: ChannelSessionId;
  user_id: ChannelUserId;
  agent_type: string;
  conversation_id?: ConversationId;
  workspace?: string;
  chatId?: string;
  created_at: number;
  lastActivity: number;
}

File: crates/backend/nomifun-db/src/models/channel.rs
use nomifun_common::TimestampMs;
use serde::{Deserialize, Serialize};

/// Row mapping for the `channel_plugins` table.
///
/// One row per connected bot — multiple rows may share the same platform
/// `type` (legacy rows keep `id == type`). The `config` column holds an
/// encrypted JSON blob containing credentials and options.
#[derive(Debug, Clone, Serialize, Deserialize, sqlx::FromRow)]
pub struct ChannelPluginRow {
    pub id: String,
    /// Platform type (telegram, lark, dingtalk, weixin, slack, discord).
    #[sqlx(rename = "type")]
    pub r#type: String,
    pub name: String,
    pub enabled: bool,
    /// JSON blob: `{ credentials, config }`. Stored encrypted at rest.
    pub config: String,
    pub status: Option<String>,
    pub last_connected: Option<TimestampMs>,
    /// Companion bound to this bot. UNIQUE(type, bot_key) guarantees a bot is
    /// never bound to more than one companion.
    pub companion_id: Option<String>,
    /// 对外伙伴 (public agent) bound to this bot. Row-level mutually exclusive
    /// with `companion_id`: a bot serves EITHER a companion OR a public agent OR
    /// nothing, never both (enforced in the repository/manager layer).
    pub public_agent_id: Option<String>,
    /// Platform-level bot identity (lark app_id, telegram bot id, ...),
    /// extracted from credentials on enable/restore.
    pub bot_key: Option<String>,
    pub created_at: TimestampMs,
    pub updated_at: TimestampMs,
}

/// Row mapping for the `channel_users` table.
///
/// Represents an IM user authorized to chat with the Agent.
/// UNIQUE constraint on (platform_user_id, platform_type).
#[derive(Debug, Clone, Serialize, Deserialize, sqlx::FromRow)]
pub struct ChannelUserRow {
    pub id: String,
    pub platform_user_id: String,
    pub platform_type: String,
    /// The `channel_plugins` row (bot) this authorization belongs to.
    /// `None` only for legacy rows the 004 migration could not backfill.
    pub channel_id: Option<String>,
    pub display_name: Option<String>,
    pub authorized_at: TimestampMs,
    pub last_active: Option<TimestampMs>,
    pub session_id: Option<String>,
}

/// Row mapping for the `channel_sessions` table.
///
/// Per-chat session linking an authorized user to a conversation.
/// FK: user_id → channel_users(id) ON DELETE CASCADE.
/// FK: conversation_id → conversations(id) ON DELETE SET NULL.
#[derive(Debug, Clone, Serialize, Deserialize, sqlx::FromRow)]
pub struct ChannelSessionRow {
    pub id: String,
    pub user_id: String,
    pub agent_type: String,
    pub conversation_id: Option<String>,
    pub workspace: Option<String>,
    pub chat_id: Option<String>,
    /// The `channel_plugins` row this session arrived through. Two bots
    /// in the same chat get isolated sessions.
    pub channel_id: Option<String>,
    pub created_at: TimestampMs,
    pub last_activity: TimestampMs,
}

/// Row mapping for the `channel_pairing_codes` table.
///
/// 6-digit pairing code with 10-minute expiry. Status transitions:
/// pending → approved | rejected | expired.
#[derive(Debug, Clone, Serialize, Deserialize, sqlx::FromRow)]
pub struct ChannelPairingCodeRow {
    pub code: String,
    pub platform_user_id: String,
    pub platform_type: String,
    /// The bot channel this pairing was initiated through.
    pub channel_id: Option<String>,
    pub display_name: Option<String>,
    pub requested_at: TimestampMs,
    pub expires_at: TimestampMs,
    pub status: String,
}
